File: src/agent/src/Safir.Agent/Services/DataDirectoryWatcher.cs
using System;
using System.IO;
using System.Reactive.Linq;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using Safir.Agent.Configuration;
using Safir.Agent.Domain;

namespace Safir.Agent.Services
{
    internal sealed class DataDirectoryWatcher : IHostedService, IFileWatcher
    {
        private static readonly IObservable<FileSystemEventArgs> _emptyObservable =
            Observable.Empty<FileSystemEventArgs>();

        private readonly IOptions<AgentOptions> _options;
        private readonly IDirectory _directory;
        private readonly ILogger<DataDirectoryWatcher> _logger;
        private FileSystemWatcher? _fileWatcher;

        public DataDirectoryWatcher(
            IOptions<AgentOptions> options,
            IDirectory directory,
            ILogger<DataDirectoryWatcher> logger)
        {
            _options = options ?? throw new ArgumentNullException(nameof(options));
            _directory = directory ?? throw new ArgumentNullException(nameof(directory));
            _logger = logger;
        }

        public IObservable<FileSystemEventArgs> Created { get; private set; } = _emptyObservable;

        public IObservable<FileSystemEventArgs> Changed { get; private set; } = _emptyObservable;

        public IObservable<FileSystemEventArgs> Deleted { get; private set; } = _emptyObservable;

        public IObservable<RenamedEventArgs> Renamed { get; private set; } = Observable.Empty<RenamedEventArgs>();

        public IObservable<ErrorEventArgs> Error { get; private set; } = Observable.Empty<ErrorEventArgs>();

        public Task StartAsync(CancellationToken cancellationToken)
        {
            _logger.LogDebug("Starting data directory watcher");

            var root = _options.Value.DataDirectory;
            if (string.IsNullOrWhiteSpace(root))
            {
                _logger.LogInformation("No data directory set");
                return Task.CompletedTask;
            }

            if (!_directory.Exists(root))
            {
                _logger.LogError("Data directory does not exist");
                return Task.CompletedTask;
            }

            _logger.LogTrace("Creating filesystem watcher");
            _fileWatcher = new FileSystemWatcher(root) {
                EnableRaisingEvents = true,
                IncludeSubdirectories = true,
            };

            _logger.LogTrace("Assigning filesystem watcher event handlers");
            CreateObservablesFromEvents(this, _fileWatcher);

            _logger.LogTrace("Finishing data directory watcher start");
            return Task.CompletedTask;
        }

        public Task StopAsync(CancellationToken cancellationToken)
        {
            _logger.LogInformation("Stopping data directory watcher");

            if (_fileWatcher == null)
            {
                _logger.LogTrace("No file watcher created, returning");
                return Task.CompletedTask;
            }

            _logger.LogTrace("Disposing file watcher");
            _fileWatcher.Dispose();

            _logger.LogTrace("Finishing data directory watcher stop");
            return Task.CompletedTask;
        }

        private static void CreateObservablesFromEvents(DataDirectoryWatcher service, FileSystemWatcher watcher)
        {
            service.Created = Observable.FromEventPattern<FileSystemEventHandler, FileSystemEventArgs>(
                    x => watcher.Created += x,
                    x => watcher.Created -= x)
                .Select(x => x.EventArgs);

            service.Changed = Observable.FromEventPattern<FileSystemEventHandler, FileSystemEventArgs>(
                    x => watcher.Changed += x,
                    x => watcher.Changed -= x)
                .Select(x => x.EventArgs);

            service.Deleted = Observable.FromEventPattern<FileSystemEventHandler, FileSystemEventArgs>(
                    x => watcher.Deleted += x,
                    x => watcher.Deleted -= x)
                .Select(x => x.EventArgs);

            service.Renamed = Observable.FromEventPattern<RenamedEventHandler, RenamedEventArgs>(
                    x => watcher.Renamed += x,
                    x => watcher.Renamed -= x)
                .Select(x => x.EventArgs);

            service.Error = Observable.FromEventPattern<ErrorEventHandler, ErrorEventArgs>(
                    x => watcher.Error += x,
                    x => watcher.Error -= x)
                .Select(x => x.EventArgs);
        }
    }
}

File: src/agent/src/Safir.Agent/Services/DataDirectoryWatcher.fs
namespace Safir.Agent.Services

open System.IO
open System.IO.Abstractions
open System.Reactive.Linq
open System.Threading.Tasks
open FSharp.Core
open Microsoft.Extensions.Hosting
open Microsoft.Extensions.Logging
open Microsoft.Extensions.Options
open Safir.Agent
open Safir.Agent.Configuration
open Safir.FSharp.Common

type DataDirectoryWatcher(options: IOptions<AgentOptions>, directory: IDirectory, logger: ILogger<DataDirectoryWatcher>) =
    let createWatcher =
        result {
            let! root =
                options.Value.DataDirectory
                |> Validation.dataDirectory directory.Exists

            logger.LogTrace("Creating filesystem watcher for {Root}", root)
            return Some(new FileSystemWatcher(root, EnableRaisingEvents = true, IncludeSubdirectories = true))
        }
        |> onError (fun e -> logger.LogInformation(Validation.getMessage e))
        |> defaultValue None

    let watcher: Lazy<FileSystemWatcher option> =
        lazy createWatcher

    interface IFileWatcher with
        member this.Changed =
            match watcher.Value with
            | None -> Observable.Empty<FileSystemEventArgs>()
            | Some w -> w.Changed

        member this.Created =
            match watcher.Value with
            | None -> Observable.Empty<FileSystemEventArgs>()
            | Some w -> w.Created

        member this.Deleted =
            match watcher.Value with
            | None -> Observable.Empty<FileSystemEventArgs>()
            | Some w -> w.Deleted

        member this.Error =
            match watcher.Value with
            | None -> Observable.Empty<ErrorEventArgs>()
            | Some w -> w.Error

        member this.Renamed =
            match watcher.Value with
            | None -> Observable.Empty<RenamedEventArgs>()
            | Some w -> w.Renamed

    interface IHostedService with
        member this.StartAsync _ =
            do logger.LogTrace("Ensuring watcher created")
            watcher.Value |> ignore
            Task.CompletedTask

        member this.StopAsync _ =
            do logger.LogInformation("Stopping data directory watcher")

            match watcher.Value with
            | None -> do logger.LogTrace("No file watcher to dispose")
            | Some w ->
                do logger.LogTrace("Disposing file watcher")
                do w.Dispose()

            do logger.LogTrace("Finishing data directory watcher cleanup")
            Task.CompletedTask
